product_variant_default_code/models/product.py

```diff
@@ -0,0 +1,196 @@
+
+
+def extract_token(s):
+    pattern = re.compile(r'\[([^\]]+?)\]')
+    return set(pattern.findall(s))
+
+
+def sanitize_reference_mask(product, mask):
+    tokens = extract_token(mask)
+    attribute_names = set()
+    for line in product.attribute_line_ids:
+        attribute_names.add(line.attribute_id.name)
+    if not tokens.issubset(attribute_names):
+        raise UserError(_('Found unrecognized attribute name in "Variant '
+                          'Reference Mask"'))
+
+
+def get_rendered_default_code(product, mask):
+    product_attrs = defaultdict(str)
+    reference_mask = ReferenceMask(mask)
+    for value in product.attribute_value_ids:
+        if value.attribute_id.code:
+            product_attrs[value.attribute_id.name] += value.attribute_id.code
+        if value.code:
+            product_attrs[value.attribute_id.name] += value.code
+    all_attrs = extract_token(mask)
+    missing_attrs = all_attrs - set(product_attrs.keys())
+    missing = dict.fromkeys(missing_attrs, PLACE_HOLDER_4_MISSING_VALUE)
+    product_attrs.update(missing)
+    default_code = reference_mask.safe_substitute(product_attrs)
+    return default_code
+
+
+def render_default_code(product, mask):
+    product.default_code = get_rendered_default_code(product, mask)
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    reference_mask = fields.Char(
+        string='Variant reference mask',
+        help='Reference mask for building internal references of a '
+             'variant generated from this template.\n'
+
+             'Example:\n'
+             'A product named ABC with 2 attributes: Size and Color:\n'
+
+             'Product: ABC\n'
+             'Color: Red(r), Yellow(y), Black(b)  #Red, Yellow, Black are '
+             'the attribute value, `r`, `y`, `b` are the corresponding code\n'
+             'Size: L (l), XL(x)\n'
+
+             'When setting Variant reference mask to `[Color]-[Size]`, the '
+             'default code on the variants will be something like `r-l` '
+             '`b-l` `r-x` ...\n'
+
+             'If you like, You can even have the attribute name appear more'
+             ' than once in the mask. Such as , `fancyA/[Size]~[Color]~[Size]`'
+             ' When saved, the default code on variants will be something like'
+             ' `fancyA/l~r~l` (for variant with Color "Red" and Size "L") '
+             '`fancyA/x~y~x` (for variant with Color "Yellow" and Size "XL")\n'
+
+             'Note: make sure characters "[,]" do not appear in your '
+             'attribute name')
+
+    @api.model
+    def create(self, vals):
+        product = self.new(vals)
+        if not vals.get('reference_mask') and product.attribute_line_ids:
+            attribute_names = []
+            for line in product.attribute_line_ids:
+                attribute_names.append("[{}]".format(line.attribute_id.name))
+            default_mask = DEFAULT_REFERENCE_SEPARATOR.join(attribute_names)
```

Comment:
@pedrobaeza @chienandalu 
Firstly I think default reference seperator either needs to be an empty string or configurable (or both) - the hyphen is annoying and it was something I removed in order to bring compatibility with product_code_builder.

Second, the default mask should be default_code + ''.join([attrs]).

Do those 2 things then you have identical functionality to the original product_code_builder with a lot more flexibility.

Given its a couple mins work to bring both these modules into line, we should do it before merging IMO.  

\cc @sebastienbeau

Reply:
@pedrobaeza now I remember why I ported that sequence module.  For @sebastienbeau requirement.

I think it can be resolved by 
```


if not vals.get('reference_mask') and product.attribute_line_ids:
    default_mask = self._get_default_mask(product)

...
def _get_default_mask(self, product)
    attribute_names = []
    for line in product.attribute_line_ids:
        attribute_names.append("[{}]".format(line.attribute_id.name))
    default_mask = DEFAULT_REFERENCE_SEPARATOR.join(attribute_names)
    return default_mask
````
What do you think?

Reply:
Hi, indeed, maybe it worst to merge both module now.

So we need to add the field 'prefix_code' on the product template, (the default or custom mask will only add extra char to the prefix_code to build the default_code). If the mask is empty we generate the code base on a default_function or we generate on the fly a default mask

We may add an extra group (with a option in the configuration menu) to show or not the mask on the product template

@chienandalu @pedrobaeza are you ok for merging both module

@bealdav can you propose a PR with the merge?

Reply:
Yes, I think we should work to have only one module. My colleague @chienandalu is already working for making this, so you owe us a drink!

Reply:
+1 to merge features